### PR TITLE
Add cache control to hub

### PR DESCRIPTION
**Pull request by embarnard**
This work comes out of [this BugCrowd ticket](https://tracker.bugcrowd.com/code-for-america-mbb/submissions/7b1c21b0-c571-4582-afe5-d3741735687f?duplicate[]=false&sort[]=severity-asc&substate[]=unresolved). Basically in the hub, a user could logout and then click the back button and return to the page that they were previously on. Adding this `set_cache_headers` on the hub basecontroller should clear the cache and prevent this from happening. 

References to this work came from [Arin's PR in GCF](https://github.com/codeforamerica/gcf-backend/pull/8866) and this [stackoverflow post ](https://stackoverflow.com/questions/711418/how-to-prevent-browser-page-caching-in-rails/748646#748646).

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -480,6 +480,13 @@ def wrapping_layout
     "application"
   end
 
+  def set_cache_headers
+    # prevents browser caching on pages set with before action
+    response.headers["Cache-Control"] = "no-cache, no-store"
+    response.headers["Pragma"] = "no-cache"
+    response.headers["Expires"] = "Mon, 01 Jan 1990 00:00:00 GMT"
+  end
```
Comment (bengolder) [inline app/controllers/application_controller.rb:488]: Not important, might be nice: I feel like there could be a clearer name for this method, like `set_no_cache_headers` `no_cache_headers` or `prevent_response_caching` or `do_not_cache_response`.

File: app/controllers/hub/admin/experiment_participants_controller.rb [hunk 2]
```diff
@@ -1,8 +1,6 @@
 module Hub
   module Admin
-    class ExperimentParticipantsController < ApplicationController
-      include AccessControllable
-      before_action :require_sign_in
+    class ExperimentParticipantsController < Hub::BaseController
```
Comment (bengolder) [inline app/controllers/hub/admin/experiment_participants_controller.rb:3]: Great refactor! I love this

File: app/controllers/hub/base_controller.rb
Comment (bengolder) [inline app/controllers/hub/base_controller.rb:1]: YES! This is such a great move! Love it!

File: app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb [hunk 3]
```diff
@@ -1,14 +1,10 @@
...
-      include AccessControllable
-
-      layout "hub"
-
-      before_action :require_sign_in
+    class BaseBulkActionsController < Hub::BaseController
       load_and_authorize_resource :tax_return_selection
       before_action :load_clients, :load_template_variables
       before_action :load_edit_form, only: :edit
+      layout "hub"
```
Comment (bengolder) [inline app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb:7]: Would it make sense to also move `layout "hub"` into the `Hub::BaseController` or would that cause other problems?

File: spec/features/hub/cache_control_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  context "user on profile page, logs out and clicks back button", js: true do
+    it "cache is cleared and redirects to the login page" do
+      visit hub_user_profile_path
+      click_on 'Sign out'
+      page.evaluate_script('window.history.back()')
+      expect(page).to have_text "Sign in"
+    end
+  end
```
Comment (bengolder) [inline spec/features/hub/cache_control_spec.rb:17]: Question for my own learning: Were you able to recreate the reported bug with this test before adding the no-cache headers?

File: app/controllers/hub/users/strong_passwords_controller.rb
Comment (bengolder) [inline app/controllers/hub/users/strong_passwords_controller.rb:1]: Question for my own learning: How did you figure out which controllers could inherit from the new hub base vs. which needed to be handled separately? Was it a matter of looking through all the controllers in the hub folder?

**Review by bengolder — APPROVED** [review]
This PR made my day, great refactor! I love the new base controller and the net negative number of lines! This is fantastic work!!

Approved with some questions for my own learning, and an unimportant "might be nice" suggestion.

File: app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb (hunk 3 above)
Comment (embarnard) [reply app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb:7]: Yeah I originally tried that but there were a few controllers that didn't include `layout "hub"` in the first place so I decided against it

File: spec/features/hub/cache_control_spec.rb (hunk 4 above)
Comment (embarnard) [reply spec/features/hub/cache_control_spec.rb:17]: Yes! Good question, the tests failed when I commented out the set_cache_headers method

File: app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb (hunk 3 above)
Comment (bengolder) [reply app/controllers/hub/bulk_actions/base_bulk_actions_controller.rb:7]: I wonder if we should move it in, and change the few to have `layout "application"`

**Pull request merged**